### PR TITLE
Mod:refactor to add base_registry

### PR DESCRIPTION
**Pull request by hxmhlt**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: registry/base_registry.go [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+		err := r.register(confIf)
+		if err != nil {
+			logger.Errorf("(ZkProviderRegistry)register(conf{%#v}) = error{%#v}",
+				confIf, perrors.WithStack(err))
+			flag = false
+			break
+		}
+		logger.Infof("success to re-register service :%v", confIf.Key())
+	}
+	r.facadeBasedRegistry.InitListeners()
```
Comment (pantianying) [inline registry/base_registry.go]: There is no such step in etcd logic, so you need to check it

File: registry/zookeeper/registry.go [hunk 2]
```diff
@@ -92,21 +69,16 @@ func newZkRegistry(url *common.URL) (registry.Registry, error) {
...
+		zkPath: make(map[string]int),
 	}
+	r.InitBaseRegistry(url, r)
 
 	err = zookeeper.ValidateZookeeperClient(r, zookeeper.WithZkName(RegistryZkClient))
 	if err != nil {
 		return nil, err
 	}
 
-	r.wg.Add(1)
```
Comment (pantianying) [inline registry/zookeeper/registry.go]: why delete this

**Comment by pantianying** [comment]
Can you remove the recursive logic from the zookeeper registry center in this PR，distinguish from configuration center😛😘

File: registry/zookeeper/registry.go (hunk 2 above)
Comment (hxmhlt) [reply registry/zookeeper/registry.go]: > why delete this

The logic is written in base_registry.

**Comment by hxmhlt** [comment]
> Did you test it end to end? Or run the samples?

Yes，I run it in dubbo-samples

File: registry/base_registry.go (hunk 1 above)
Comment (hxmhlt) [reply registry/base_registry.go]: > There is no such step in etcd logic, so you need to check it
Which step?

**Comment by hxmhlt** [comment]
> Can you remove the recursive logic from the zookeeper registry center in this PR，distinguish from configuration center😛😘

I think it will be another pr. We need to create a new pull request to do that.

File: registry/base_registry.go [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+	processID = ""
+	localIP   = ""
+)
+
+func init() {
+	processID = fmt.Sprintf("%d", os.Getpid())
+	localIP, _ = gxnet.GetLocalIP()
+}
+
+type FacadeBasedRegistry interface {
```
Comment (zouyx) [inline registry/base_registry.go]: Should you add some comment for this interface

File: registry/base_registry.go [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+	Registry
+	CreatePath(string) error
+	DoRegister(string, string) error
+	DoSubscribe(conf *common.URL) (Listener, error)
+	CloseAndNilClient()
+	CloseListener()
+	InitListeners()
+}
+
+type BaseRegistry struct {
```
Comment (zouyx) [inline registry/base_registry.go]: Should you add some comment for this class

File: registry/base_registry.go (hunk 3 above)
Comment (hxmhlt) [reply registry/base_registry.go]: done

File: registry/base_registry.go (hunk 4 above)
Comment (hxmhlt) [reply registry/base_registry.go]: done

File: registry/base_registry.go [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+	facadeBasedRegistry FacadeBasedRegistry
+	*common.URL
+	birth int64          // time of file birth, seconds since Epoch; 0 if unknown
+	wg    sync.WaitGroup // wg+done for zk restart
+	done  chan struct{}
+
+	cltLock  sync.Mutex
+	services map[string]common.URL // service name + protocol -> service config
+}
+
```
Comment (AlexStocks) [inline registry/base_registry.go]: A comment pls.

File: registry/base_registry.go [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+func (r *BaseRegistry) InitBaseRegistry(url *common.URL, facadeRegistry FacadeBasedRegistry) Registry {
+	r.URL = url
+	r.birth = time.Now().UnixNano()
+	r.done = make(chan struct{})
+	r.services = make(map[string]common.URL)
+	r.facadeBasedRegistry = facadeRegistry
+	r.wg.Add(1)
+	return r
+}
+
```
Comment (AlexStocks) [inline registry/base_registry.go]: comment

File: registry/base_registry.go [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+func (r *BaseRegistry) GetUrl() common.URL {
+	return *r.URL
+}
+func (r *BaseRegistry) Destroy() {
```
Comment (AlexStocks) [inline registry/base_registry.go]: comment

File: registry/base_registry.go [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+	r.facadeBasedRegistry.CloseListener()
+	close(r.done)
+	r.wg.Wait()
+	r.closeRegisters()
+}
+
```
Comment (AlexStocks) [inline registry/base_registry.go]: comment and split it.

File: registry/base_registry.go [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+		logger.Debugf("(ProviderRegistry)Register(conf{%#v})", conf)
+	}
+
+	return nil
+}
+
+func (r *BaseRegistry) service(c common.URL) string {
+	return url.QueryEscape(c.Service())
+}
+
```
Comment (AlexStocks) [inline registry/base_registry.go]: comment.

File: registry/base_registry.go [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+			break
+		}
+		logger.Infof("success to re-register service :%v", confIf.Key())
+	}
+	r.facadeBasedRegistry.InitListeners()
+
+	return flag
+}
+
+func (r *BaseRegistry) register(c common.URL) error {
```
Comment (AlexStocks) [inline registry/base_registry.go]: pls split it.

File: registry/base_registry.go [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+
+func sleepWait(n int) {
+	wait := time.Duration((n + 1) * 2e8)
+	if wait > MaxWaitInterval {
+		wait = MaxWaitInterval
+	}
+	time.Sleep(wait)
+}
+
+//subscribe from registry
```
Comment (AlexStocks) [inline registry/base_registry.go]: "Subscribe" the func name as the first comment word.

File: registry/base_registry.go [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+
+func (r *BaseRegistry) closeRegisters() {
+	r.cltLock.Lock()
+	defer r.cltLock.Unlock()
+	logger.Infof("begin to close provider client")
+	// Close the old client first to close the tmp node.
+	r.facadeBasedRegistry.CloseAndNilClient()
+	r.services = nil
+}
+
```
Comment (AlexStocks) [inline registry/base_registry.go]: Add comment.

File: registry/base_registry.go [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+func (r *BaseRegistry) IsAvailable() bool {
+	select {
+	case <-r.done:
+		return false
+	default:
+		return true
+	}
+}
+
```
Comment (AlexStocks) [inline registry/base_registry.go]: add comment

File: registry/base_registry.go [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+func (r *BaseRegistry) WaitGroup() *sync.WaitGroup {
```
Comment (AlexStocks) [inline registry/base_registry.go]: Add comment.

File: registry/base_registry.go [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+	return &r.wg
+}
+
+func (r *BaseRegistry) Done() chan struct{} {
```
Comment (AlexStocks) [inline registry/base_registry.go]: Add comment.

File: registry/etcdv3/listener.go [hunk 16]
```diff
@@ -81,7 +81,7 @@ type configurationListener struct {
 // NewConfigurationListener ...
```
Comment (AlexStocks) [inline registry/etcdv3/listener.go]: pls help to change the comment.

File: registry/base_registry.go (hunk 6 above)
Comment (hxmhlt) [reply registry/base_registry.go]: It looks like reasonable to move outside so I do it

**Comment by AlexStocks** [comment]
LGTM

**Pull request merged**